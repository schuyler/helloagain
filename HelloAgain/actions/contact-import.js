'use strict'

import Contacts from 'react-native-contacts'

import {
  CONTACTS_LOADED, CONTACT_LOAD_FAILED
} from './types';

export const contactsLoaded = (contacts) => {
  return {type: CONTACTS_LOADED, contacts: contacts}
}

export const contactLoadFailed = (error) => {
  return {type: CONTACT_LOAD_FAILED, error: error}
}

export const loadNativeContacts = () => {
  // Create a Promise that gets all native contacts and resolves on success.
  const loadContacts = new Promise((resolve, reject) => {
    Contacts.getAll((err, contacts) => {
      err ? reject(err) : resolve(contacts)
    })
  })
  // Return an action thunk that:
  //   1. Loads all contacts
  //   2. On success, dispatches the `contactsLoaded` action
  //   3. Then, dispatches the `writeIDsForAllFriends` to write back any new
  //      IDs to the native contact store
  //   4. On failure, dispatch the `contactLoadFailed` action
  return dispatch => {
    return loadContacts
      .then(contacts => {
        let result = dispatch(contactsLoaded(contacts))
        console.log("contactsLoaded:", Object.values(result).length)
      })
      .then(_ => { return dispatch(writeIDsForAllFriends()) })
      .catch(error => dispatch(contactLoadFailed(error)))
  }
}

const generateURL = (item) => {
  return "helloagain://to/" + item.helloAgainID
}

// Check a single contact record to see if it needs the HelloAgain URL
// written to it, and if it does, write it.
export const writeIDtoNativeContact = (item) => {
  // Never been active? Don't touch the native contact.
  if (item.isActive === undefined) {
    return
  }
  // Are any of the contact's urlAddresses a HelloAgain address? Great. Skip
  // this one.
  let matchingURLs = (item.urlAddresses || []).filter(url => url.label == "HelloAgain")
  if (matchingURLs.length > 0) {
    return
  }
  // Otherwise, generate the URL, push it onto the urlAddresses list with the
  // right label, and then update the native contact store, with a log message
  // either way.
  let url = generateURL(item)
  item.urlAddresses.push({label: "HelloAgain", urlAddress: url})
  Contacts.updateContact(item, (err, contact) => {
    if (err) {
      console.log(`Error updating ${item.givenName} ${item.familyName} ` +
                  `with URL ${url}:`, err, contact)
    } else {
      console.log(`Wrote URL to Contacts: ${url}`)
    }
  })
}

const writeIDsForAllFriends = () => {
  // Return an action thunk that is pure side-effects -- just iterate
  // over the entire set of contacts and make sure that any once-active friends
  // that don't have HelloAgain URLs associated with them get one. This
  // shouldn't really ever happen, because the `toggleActive` action takes care
  // of the records one-by-one, but I guess it can't hurt to be certain.
  return (_, getState) => {
    const state = getState()
    const friends = Object.values(state.friends).filter(f => f.isActive !== undefined)
    friends.forEach(item => writeIDtoNativeContact(item))
  }
}
