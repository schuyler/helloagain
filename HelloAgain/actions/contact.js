'use strict'

import { TOGGLE_ACTIVE } from './types';

export const toggleActive = (friend) => {
  return { type: TOGGLE_ACTIVE, friend: friend }
}
