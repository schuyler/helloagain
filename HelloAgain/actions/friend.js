'use strict'

import {
  UPDATE_FRIEND,
  MARK_AS_CONTACTED
} from './types'

export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}

export const markAsContacted = (friend) => {
  return { type: MARK_AS_CONTACTED, friend: friend }
}
