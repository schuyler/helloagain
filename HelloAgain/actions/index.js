'use strict';

// Enumerate the dispatchable actions.
export const UPDATE_FRIEND = 'UPDATE_FRIEND'

// Helpers for dispatching actions.
export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}
