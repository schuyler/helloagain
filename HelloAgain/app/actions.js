export const IMPORT_CONTACTS = 'IMPORT_CONTACTS';

export const UPDATE_FRIEND = 'UPDATE_FRIEND';

export const FRIEND_CONTACTED = 'FRIEND_CONTACTED';

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';

export function importContacts(contactList) {
  return { type: IMPORT_CONTACTS, contactList };
}

export function updateFriend(friend) {
  return { type: UPDATE_FRIEND, friend };
}
