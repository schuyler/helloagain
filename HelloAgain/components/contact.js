'use strict';

import React, { Component } from 'react';
import {
  Text,
  View,
  StyleSheet,
  TouchableOpacity,
} from 'react-native'
import Portrait from './portrait'
import shared from './shared'

export default class Contact extends Component {
  constructor(props) {
    super(props)
  }

  render() {
    const item = this.props.item;
    return (
      <TouchableOpacity onPress={() => this.props.onPress(item)}>
        <View style={styles.contactRow}>
          <Portrait uri={item.thumbnailPath} style={styles.contactPicture} />
          <Text style={styles.contactName}>{item.givenName} {item.familyName}</Text>
          <Text style={item.isActive ? styles.contactActive : styles.contactInactive}>â˜…</Text>
        </View>
      </TouchableOpacity>
    )
  }
}

const styles = StyleSheet.create({
  contactRow: shared.row,
  contactPicture: shared.rowPicture,
  contactName: shared.rowName,
  contactInactive: {
    ...shared.rowProps,
    color: 'lightgrey'
  },
  contactActive: {
    ...shared.rowProps,
    color: 'dodgerblue'
  }
})

