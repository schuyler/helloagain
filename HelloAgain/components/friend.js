'use strict';

import React, { Component } from 'react';
import {
  Text,
  View,
  StyleSheet,
  TouchableOpacity,
  Image
} from 'react-native'
import Duration from 'durationjs'
import Portrait from './portrait'
import shared from './shared'

export default class Friend extends Component {
  constructor(props) {
    super(props)
  }

  formatDurationSince(then) {
    if (!then) {
      return "never"
    }
    let ago = true ? " ago" : ""
    let delta = Date.now() - then
    let age = new Duration(delta / 1000)
    if (age.inMonths() >= 1) {
      return Math.floor(age.inMonths()) + "m" + ago
    } else if (age.inWeeks() >= 1) {
      return Math.floor(age.inWeeks()) + "w" + ago
    } else if (age.inDays() >= 1) {
      return Math.floor(age.inDays()) + "d" + ago
    } else if (age.inHours() >= 1) {
      return Math.floor(age.inHours()) + "h" + ago
    } else if (age.inMinutes() >= 1) {
      return Math.floor(age.inMinutes()) + "m" + ago
    } else {
      return "just now"
    }
  }

  render() {
    const item = this.props.item;
    let sinceLastContact = this.formatDurationSince(item.contactedAt)
    return (
      <TouchableOpacity onPress={() => this.props.onPress(item)}>
        <View style={styles.contactRow}>
          <Portrait uri={item.thumbnailPath} style={styles.contactPicture} />
          <Text style={styles.contactName}>{item.givenName} {item.familyName}</Text>
          <Text style={styles.contactLatency}>{sinceLastContact}</Text>
        </View>
      </TouchableOpacity>
    )
  }
}

// This really wants to be refactored somehow with the contact styles
const styles = StyleSheet.create({
  contactRow: shared.row,
  contactPicture: shared.rowPicture,
  contactName: shared.rowName,
  contactLatency: {
    ...shared.rowProps,
    fontSize: shared.rowProps.fontSize - 10,
    color: "grey"
  }  
})
