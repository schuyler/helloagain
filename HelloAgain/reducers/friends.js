'use strict';

import { 
  UPDATE_FRIEND,
  TOGGLE_ACTIVE,
  CONTACTS_LOADED,
  MARK_AS_CONTACTED
} from '../actions/types';

const _id = (friend) => {
  return friend.recordID
}

const findFriend = (state, friend) => {
  return state[_id(friend)] 
}

const queueTail = (state) => {
  const ranks = Object.values(state).map((f) => {
    return f.rank || 0
  })
  return Math.max(0, ...ranks)
}

const defaultValues = (state) => {
  return {
    rank: queueTail(state) + 1,
    isActive: false
  }
}

const markContacted = (state, friend) => {
  return {
    ...friend,
    rank: queueTail(state) + 1,
    contactedAt: Date.now()
  }
}

const updateFriend = (state, update, useDefaults) => {
  let id = _id(update)
  let defaults = useDefaults ? defaultValues(state) : undefined
  let newFriend = {...defaults, ...state[id], ...update}
  return {...state, [id]: newFriend}
}

const importContacts = (state, contacts) => {
  let newState = {...state}
  contacts.forEach((item) => {
    let id = _id(item)
    newState[id] = {...state[id], ...item}
  })
  return newState
}

const friends = (state = {}, action) => {
  switch (action.type) {
    case CONTACTS_LOADED:
      return importContacts(state, action.contacts)
    case TOGGLE_ACTIVE:
      let existingFriend = findFriend(state, action.friend)
      let isActive = (existingFriend ? existingFriend.isActive : false)
      let update = {...action.friend, isActive: !isActive}
      // Set default values (if needed) when toggling a friend
      return updateFriend(state, update, true)
    case MARK_AS_CONTACTED:
      let asContacted = markContacted(state, action.friend)
      return updateFriend(state, asContacted)
    case UPDATE_FRIEND:
      return updateFriend(state, action.friend)
    default:
      return state
  }
}

export default friends
