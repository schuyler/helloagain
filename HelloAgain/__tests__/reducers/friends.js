'use strict';

import friends from '../../reducers/friends'
import * as actions from '../../actions/friend'
import * as contactActions from '../../actions/contact'

describe('friends reducer', () => {
  let bob = {name: "Bob", recordID: 50}

  it('should return an initial state', () => {
    expect(
      friends(undefined, {})
    ).toEqual({})
  })

  describe('updateFriend action', () => {
    let initialState = friends(undefined, actions.updateFriend(bob))
    it('should add a friend', () => {
      expect(initialState).toEqual({50: bob})
    })

    it('should not duplicate an existing friend', () => {
      expect(
        friends(initialState, actions.updateFriend(bob))
      ).toEqual(initialState)
    })

    it("should merge a friend's existing record", () => {
      let newFact = {recordID: 50, hobby: "fishing"}
      let newState = friends(initialState, actions.updateFriend(newFact))
      expect(newState).toMatchObject({50: {...bob, ...newFact}})
    })

    let jim = {name: "Jim", recordID: 42};
    it('should add a new friend at the front', () => {
      expect(
        friends(initialState, actions.updateFriend(jim))
      ).toEqual({50: bob, 42: jim})
    })
  })

  describe('toggleActive action', () => {
    it('should toggle a new friend active', () => {
      let state = friends(undefined, contactActions.toggleActive(bob))
      expect(state).toMatchSnapshot()
    })

    it('should toggle an existing friend (in)active', () => {
      let state = friends(undefined, actions.updateFriend({...bob, isActive: true}))

      // Now Bob is inactive
      state = friends(state, contactActions.toggleActive(bob))
      expect(state).toMatchSnapshot()

      // Now Bob is active
      state = friends(state, contactActions.toggleActive(bob))
      expect(state).toMatchSnapshot()
    })
  })
})
