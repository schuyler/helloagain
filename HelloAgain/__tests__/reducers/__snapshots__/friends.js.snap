exports[`friends reducer toggleActive action should toggle a new friend active 1`] = `
Object {
  "50": Object {
    "isActive": true,
    "name": "Bob",
    "recordID": 50,
  },
}
`;

exports[`friends reducer toggleActive action should toggle an existing friend (in)active 1`] = `
Object {
  "50": Object {
    "isActive": false,
    "name": "Bob",
    "recordID": 50,
  },
}
`;

exports[`friends reducer toggleActive action should toggle an existing friend (in)active 2`] = `
Object {
  "50": Object {
    "isActive": true,
    "name": "Bob",
    "recordID": 50,
  },
}
`;
